[FEATURE] Add TypoScript provider for sites and sets (#1240)
TYPO3-Documentation/TYPO3CMS-Reference-Typoscript: Documentation/UsingSetting/AddTypoScriptWithExtensions.rst

Add TypoScript in your extension
================================

..  versionchanged:: 13.1

>     TypoScript on a per site basis can now be included via

Reviewer: ```suggestion
    TypoScript on a per-site basis can now be included via
```

    :ref:`sites and sets <typoscript-site-sets>`.

.. note::

   This part is written for extension developers.

.. index:: TypoScript in extensions; File locations

.. _extdev-add-typoscript-extension:

Create TypoScript files in your extension
=========================================

TypoScript files **must** have the ending :file:`.typoscript`.

> They are located in :path:`Configuration/Sets/MySet` within your

Reviewer: Hhm but having TypoScript outside the "sets" scope is still possible; isn't the section below meant for the new placement syntax, and we shouldn't change the old?

Reviewer 2: Yup, the old location is still possible. The site sets is a new way when using - well - site sets. You can still go the old way without site sets.

Author: Starting with TYPO3 v13 not using site sets is not recommended anymore. You can still put your files to be included in other directions

Reviewer 2: Then this should be mentioned. Where is the source for the recommendation? How would you configure TypoScript for all sites in your installation?

and :ref:`how to provide TypoScript for both TYPO3 v13 and v12 <extdev-add-typoscript-sets-v12>`.

*   :file:`constants.typoscript` contains the constants
*   :file:`setup.typoscript` contains the TypoScript setup

.. _extdev-add-typoscript-sets:

TypoScript provided as site set (only TYPO3 v13.1 and above)
============================================================

The file structure of the extension could, for example look like this:

..  directory-tree::
    :show-file-icons: true

    *   Classes

        *   ...

    *   Configuration

        *   Sets

            *   MyExtension

                *   :ref:`config.yaml <extdev-add-typoscript-sets-main>`
                *   :ref:`constants.typoscript <extdev-add-typoscript-sets-typoscript>`
                *   :ref:`setup.typoscript <extdev-add-typoscript-sets-typoscript>`

            *   MyExtensionWithACoolFeature

                *   :ref:`config.yaml <extdev-add-typoscript-sets-feature>`
                *   :ref:`setup.typoscript <extdev-add-typoscript-sets-typoscript>`

    *   Resources

        *   ...

    *   composer.json
    *   ...

> With the extensions TypoScript residing in :file:`EXT:my_extension/Configuration/TypoScript`

Reviewer: ```suggestion
With the extension's TypoScript residing in :file:`EXT:my_extension/Configuration/TypoScript`
```

You mean in Configuration/Sets/MySet though!? The description above already shows a file structure for Sets. Is the intention here how to move from Configuration/ to Sets?

I don't really understand (and maybe I'm a good target audience, since TypoScript and Sets are complicated to me)

> With the extension's TypoScript residing in :file:`EXT:my_extension/Configuration/TypoScript`
> and the TypoScript for some optional feature in
> :file:`EXT:my_extension/Configuration/TypoScript/SomeCoolFeature`. Let us assume, that the
> optional feature depends on the main TypoScript.

Author: ```suggestion
With the extension's TypoScript residing in :file:`EXT:my_extension/Configuration/Sets/MyExtension`
and the TypoScript for some optional feature in
:file:`EXT:my_extension/Configuration/Sets/MyExtensionWithACoolFeature`. Let us assume, that the
optional feature depends on the main TypoScript.
```

TYPO3 v12.4 and v13, you can provide site sets but still support including
the TypoScript in the :sql:`sys_template` record via `static_file_include`'s.

The files in the sets are the same as in the
:ref:`example for TYPO3 v13 only <extdev-add-typoscript-sets>`.

The extended file structure of the extension could, for example look like this:

..  directory-tree::
    :show-file-icons: true

    *   Classes

        *   ...

    *   Configuration

        *   Sets

            *   MyExtension

                *   :ref:`config.yaml <extdev-add-typoscript-sets-main>`
                *   :ref:`constants.typoscript <extdev-add-typoscript-sets-typoscript>`
                *   :ref:`setup.typoscript <extdev-add-typoscript-sets-typoscript>`

            *   MyExtensionWithACoolFeature

                *   :ref:`config.yaml <extdev-add-typoscript-sets-feature>`
                *   :ref:`setup.typoscript <extdev-add-typoscript-sets-typoscript>`

        *   TCA

            *   Overrides

                sys_template.php

    *   Resources

        *   ...

    *   composer.json
    *   ...

.. _extdev-add-typoscript-sets-v12-static_includes:
.. _extdev-static-includes:

> Only for TYPO3 v12.4: Make TypoScript available for static includes

Reviewer: ```suggestion
Only when supporting TYPO3 v12.4: Make TypoScript available for static includes
```

> Only for TYPO3 v12.4: Make TypoScript available for static includes

Reviewer 2: Only v12.4 is wrong, you can still go the old way with static includes (as sys_template records are still available and considered). The site sets are an additional way to configure TypoScript.

Author: Using static includes is not recommended anymore for TYPO3 v13.

-------------------------------------------------------------------

> Make TypoScript available for static includes (Only for TYPO3 v12.4):

Reviewer: ```suggestion
Make TypoScript available for static includes (only needed if your extensions aims to support TYPO3 v12.4):
```
